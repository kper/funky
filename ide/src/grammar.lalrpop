use std::str::FromStr;
use crate::ir::ast::*;

grammar;

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

pub Program : Program = {
   <functions:Function*> => Program {
       functions: functions
   } 
};

pub Function : Function = {
    "define" <name:Id> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Id> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    }
};

pub Instruction : Instruction = {
    "BLOCK" <n:Id> => Instruction::Block(n),
    <dest:Reg> "=" <src:Reg> => Instruction::Assign(dest, src),
    <dest:Reg> "=" <n:Num> => Instruction::Const(dest, n),
    <dest:Reg> "=" "UNKNOWN" => Instruction::Unknown(dest),
    <dest:Reg> "=" "op" <src:Reg> => Instruction::Unop(dest, src),
    <dest:Reg> "=" <src1:Reg> "op" <src2:Reg> => Instruction::BinOp(dest, src1, src2),
    <dest:Reg> "=" "phi" <src1:Reg> <src2:Reg> => Instruction::Phi(dest, src1, src2),
    <dest:Reg> "=" "STORE" <src:Reg> "TO" <offset:Reg> "ALIGN" <align:Num> <size:Num> => Instruction::Store,
    "GOTO" <n:Id> => Instruction::Jump(n),
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> "ELSE" "GOTO" <n2:Id> => {
        Instruction::Conditional(reg, vec![n, n2])
    },
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> => {
        Instruction::Conditional(reg, vec![n])
    },
    "TABLE GOTO" <n:Id*> "ELSE" "GOTO" <n2:Id> => {
        let mut n = n.clone();
        n.push(n2);
        Instruction::Table(n)
    },
    "KILL" <reg:Reg> => Instruction::Kill(reg),
    "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "RETURN" <n:Reg*> ";" => {
        Instruction::Return(n)
    },
    "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, vec![]),
    <dest:Reg*> "<-" "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, dest),
};

Id : String = {
    <n:Num> => (n as usize).to_string()
};

pub Reg: String = {
    "%" <n:Num> => format!("%{}", n),
    "%" <n:Name> => format!("%{}", n),
};

pub Num: f64 = <s:r"[+-]?[0-9]*[.,]?[0-9]+"> => f64::from_str(s).unwrap();
pub Name: String = <s:r"[a-zA-Z]+"> => s.to_string();