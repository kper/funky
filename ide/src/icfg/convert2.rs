use crate::icfg::graph2::*;
use crate::ir::ast::Function as AstFunction;
use crate::ir::ast::Instruction;
/// This module is responsible to parse
/// the webassembly AST to a graph
use crate::{counter::Counter, solver::Request};
use anyhow::{bail, Context, Result};
use std::collections::VecDeque;

use log::debug;

use std::collections::HashMap;

use crate::ir::ast::Program;

use crate::icfg::graph2::Edge;

#[derive(Debug)]
pub struct ConvertSummary {
    block_counter: Counter,
}

impl ConvertSummary {
    pub fn new() -> Self {
        Self {
            block_counter: Counter::default(),
        }
    }

    pub fn visit(&mut self, prog: &Program, req: &Request) -> Result<Graph> {
        let mut graph = Graph::new();

        self.tabulate(&mut graph, prog, req)?;

        Ok(graph)
    }

    /// computes all intraprocedural edges
    fn flow(
        &self,
        function: &AstFunction,
        graph: &mut Graph,
        pc: usize,
        variable: &String,
    ) -> Result<Vec<Edge>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut edges = Vec::new();

        let instructions = &function.instructions;

        let instruction = instructions.get(pc).context("Cannot find instr")?;
        debug!("Next instruction is {:?}", instruction);

        //graph.add_statement(function, instruction, pc + 1)?;
        let before: Vec<Fact> = graph
            .get_facts_at(&function.name, pc)?
            .into_iter()
            .filter(|x| &x.belongs_to_var == variable)
            .cloned()
            .collect();

        debug!("Facts before statement {}", before.len());

        //graph.add_statement(function, format!("{:?} ({})", instruction, pc + 1), pc + 1)?;

        let after: Vec<_> = graph
            .get_facts_at(&function.name, pc + 1)?
            .into_iter()
            .filter(|x| &x.belongs_to_var == variable)
            .collect();

        debug!("Facts after statement {}", after.len());

        for (b, a) in before.into_iter().zip(after) {
            edges.push(Edge::Normal {
                from: b,
                to: a.clone().clone(),
                curved: false,
            });
        }

        Ok(edges)
    }

    /// Computes call-to-start edges
    fn pass_args(
        &self,
        program: &Program,
        caller_function: &AstFunction,
        callee: &String,
        params: &Vec<String>,
        dests: &Vec<String>,
        graph: &mut Graph,
        current_pc: usize,
    ) -> Result<Vec<Edge>> {
        let callee_function = program
            .functions
            .iter()
            .find(|x| &x.name == callee)
            .context("Cannot find function")?;

        let facts = graph.init_function(callee_function)?;
        let caller_facts = graph.get_facts_at(&caller_function.name, current_pc)?;

        let mut edges = Vec::new();
        for (caller, callee) in caller_facts.iter().zip(facts) {
            edges.push(Edge::Call {
                from: caller.clone().clone(),
                to: callee,
            })
        }

        Ok(edges)
    }

    /// Computes exit-to-return edges
    fn return_val(
        &self,
        caller_function: &String,
        callee_function: &String,
        caller_pc: usize,
        callee_pc: usize,
        //dest: &Vec<String>,
        caller_instructions: &Vec<Instruction>,
        graph: &mut Graph,
    ) -> Result<Vec<Edge>> {
        debug!("Trying to compute return_val");
        debug!("Caller: {} ({})", caller_function, caller_pc);
        debug!("Callee: {} ({})", callee_function, callee_pc);

        let dest = match caller_instructions.get(caller_pc).as_ref() {
            Some(Instruction::Call(_, _params, dest)) => dest,
            Some(x) => bail!("Wrong instruction passed to return val. Found {:?}", x),
            None => bail!("Cannot find instruction while trying to compute exit-to-return edges"),
        };

        let caller_facts = graph.get_facts_at(caller_function, caller_pc)?;
        let callee_facts = graph.get_facts_at(callee_function, callee_pc)?;

        debug!("caller_facts {:#?}", caller_facts);
        debug!("callee_facts {:#?}", callee_facts);

        // Generate edges when for all dest + taut

        let mut edges = Vec::new();

        // taut
        /*edges.push(Edge::CallToReturn {
            from: caller_facts.get(0).unwrap().clone().clone(),
            to: callee_facts.get(0).unwrap().clone().clone(),
        });*/

        debug!("=> dest {:?}", dest);

        /*
        for (i, d) in dest.iter().enumerate() {
            let caller_fact = caller_facts
                .iter()
                .find(|x| &x.belongs_to_var == d)
                .context("Cannot find var for caller_fact")?;
            let callee_fact = callee_facts
                .get(i + 1)
                .context("Cannot find var for callee_fact")?;

            edges.push(Edge::CallToReturn {
                from: caller_fact.clone().clone(),
                to: callee_fact.clone().clone(),
            });
        }*/

        let mut index = 0;
        for callee_fact in callee_facts.iter() {
            let caller_fact = caller_facts
                .get(index)
                .context("Cannot find var for caller_fact")?;

            if caller_fact.var_is_taut || dest.contains(&caller_fact.belongs_to_var) {
                edges.push(Edge::CallToReturn {
                    from: caller_fact.clone().clone(),
                    to: callee_fact.clone().clone(),
                });

                index += 1;
            }
        }

        Ok(edges)
    }

    /// Computes call-to-return
    fn call_flow(
        &self,
        program: &Program,
        caller_function: &AstFunction,
        callee: &String,
        params: &Vec<String>,
        dests: &Vec<String>,
        graph: &mut Graph,
        pc: usize,
    ) -> Result<Vec<Edge>> {
        debug!("Generating call-to-return edges for {}", callee);

        /* 
        graph.add_statement(
            caller_function,
            format!("{:?}", "call"),
            pc + 1,
            &"taut".to_string(),
        )?;*/
        //for dest in dests.iter() {
            //graph.add_statement(caller_function, format!("{:?}", "call"), pc)?;
            //graph.add_statement(caller_function, format!("{:?}", "call"), pc + 1)?;
        //}

        let before: Vec<_> = graph
            .get_facts_at(&caller_function.name, pc)?
            .into_iter()
            .map(|x| x.clone())
            .collect();
        debug!("Facts before statement {}", before.len());

        let after = graph.get_facts_at(&caller_function.name, pc + 1)?;

        debug!("Facts after statement {}", after.len());

        let after: Vec<_> = after
            .into_iter()
            .filter(|x| !dests.contains(&x.belongs_to_var))
            .collect();

        debug!("Facts after statement without dests {}", after.len());

        debug!("before {:#?}", before);
        debug!("after {:#?}", after);

        let mut edges = Vec::with_capacity(after.len());
        for fact in after {
            let b = before
                .iter()
                .find(|x| x.belongs_to_var == fact.belongs_to_var)
                .context("Variable mismatch.")?;

            edges.push(Edge::CallToReturn {
                from: b.clone().clone(),
                to: fact.clone(),
            });
        }

        Ok(edges)
    }

    fn tabulate(&mut self, mut graph: &mut Graph, prog: &Program, req: &Request) -> Result<()> {
        debug!("Convert intermediate repr to graph");

        let function = prog
            .functions
            .iter()
            .find(|x| x.name == req.function)
            .context("Cannot find function")?;

        if graph.is_function_defined(&function.name) {
            debug!("==> Function was already summarised.");
            return Ok(());
        }

        let facts = graph.init_function(&function)?;
        let init = facts.get(0).unwrap().clone();

        let mut path_edge = Vec::new();
        let mut worklist = VecDeque::new();
        let mut summary_edge = Vec::new();

        path_edge.push(Edge::Path {
            from: init.clone(),
            to: init.clone(),
        });
        worklist.push_back(Edge::Path {
            from: init.clone(),
            to: init.clone(),
        });

        self.forward(
            &prog,
            &function,
            &mut path_edge,
            &mut worklist,
            &mut summary_edge,
            &mut graph,
        )?;

        Ok(())
    }

    fn propagate(
        &self,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        e: Edge,
    ) -> Result<()> {
        let f = path_edge.iter().rev().find(|x| *x == &e);

        if f.is_none() {
            path_edge.push(e.clone());
            worklist.push_back(e);
        }

        Ok(())
    }

    fn forward(
        &self,
        program: &Program,
        function: &AstFunction,
        path_edge: &mut Vec<Edge>,
        worklist: &mut VecDeque<Edge>,
        summary_edge: &mut Vec<Edge>,
        graph: &mut Graph,
    ) -> Result<()> {
        let mut end_summary: HashMap<(String, usize, String), Vec<Fact>> = HashMap::new();
        let mut incoming: HashMap<(String, usize, String), Vec<Fact>> = HashMap::new();

        while let Some(edge) = worklist.pop_front() {
            debug!("Popping edge from worklist {:#?}", edge);

            let d1 = edge.get_from();
            let d2 = edge.to();

            let pc = edge.to().pc;
            debug!("Instruction pointer is {}", pc);

            let instructions = &program
                .functions
                .iter()
                .find(|x| x.name == d2.function)
                .context("Cannot find function")?
                .instructions;
            let n = instructions.get(pc);
            debug!("=> Instruction {:?}", n);

            if let Some(n) = n {
                match n {
                    Instruction::Call(callee, params, dest) => {
                        let call_edges =
                            self.pass_args(program, function, callee, params, dest, graph, d2.pc)?;

                        for d3 in call_edges.into_iter() {
                            debug!("d3 {:?}", d3);

                            self.propagate(
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d3.to().clone(),
                                    to: d3.to().clone(),
                                },
                            )?; //self loop

                            debug!(
                                "Propagate {:?}",
                                Edge::Path {
                                    from: d3.to().clone(),
                                    to: d3.to().clone(),
                                }
                            );

                            //Add incoming

                            if let Some(incoming) = incoming.get_mut(&(
                                d3.to().function.clone(),
                                d3.to().pc,
                                d3.to().belongs_to_var.clone(),
                            )) {
                                incoming.push(d2.clone());
                            } else {
                                incoming.insert(
                                    (
                                        d3.to().function.clone(),
                                        d3.to().pc,
                                        d3.to().belongs_to_var.clone(),
                                    ),
                                    vec![d2.clone()],
                                );
                            }

                            debug!("Incoming in call {:#?}", incoming);

                            if let Some(end_summary) = end_summary.get(&(
                                d3.to().function.clone(),
                                d3.to().pc,
                                d3.to().belongs_to_var.clone(),
                            )) {
                                for d4 in end_summary.iter() {
                                    for d5 in self.return_val(
                                        &function.name,
                                        &d4.function,
                                        d2.pc,
                                        d4.pc,
                                        &instructions,
                                        graph,
                                    )? {
                                        summary_edge.push(Edge::Summary {
                                            from: d2.clone(),
                                            to: d5.get_from().clone(), //return_site is d5.get_from?
                                        });
                                    }
                                }
                            }

                            debug!("End summary {:#?}", end_summary);
                        }

                        let call_flow =
                            self.call_flow(program, function, callee, params, dest, graph, pc)?;

                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();

                        //TODO `or` with overwritten value
                        for d3 in call_flow {
                            let taut = graph.get_taut(&d3.get_from().function).unwrap().clone();
                            self.propagate(
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: taut,
                                    to: d3.to().clone(),
                                },
                            )?; // adding edges to return site of caller from d1
                        }
                    }
                    _ => {
                        let new_function = program
                            .functions
                            .iter()
                            .find(|x| x.name == d2.function)
                            .unwrap();
                        for f in self
                            .flow(&new_function, graph, d2.pc, &d2.belongs_to_var)?
                            .iter()
                        {
                            debug!("Normal flow {:?}", f);
                            let to = f.to();
                            self.propagate(
                                path_edge,
                                worklist,
                                Edge::Path {
                                    from: d1.clone(),
                                    to: to.clone(),
                                },
                            )?;
                        }
                    }
                }
            } else {
                // this is E_p
                debug!("=> Reached end of procedure");

                assert_eq!(d1.function, d2.function);

                // Summary
                if let Some(end_summary) =
                    end_summary.get_mut(&(d1.function.clone(), d1.pc, d1.belongs_to_var.clone()))
                {
                    let facts = graph
                        .get_facts_at(&d2.function.clone(), d2.pc)?
                        .into_iter()
                        .map(|x| x.clone());
                    end_summary.extend(facts);
                } else {
                    let facts = graph
                        .get_facts_at(&d2.function.clone(), d2.pc)?
                        .into_iter()
                        .map(|x| x.clone())
                        .collect();
                    end_summary.insert(
                        (d1.function.clone(), d1.pc, d1.belongs_to_var.clone()),
                        facts,
                    );
                }

                debug!("End Summary {:#?}", end_summary);

                // Incoming
                if let Some(incoming) =
                    incoming.get_mut(&(d1.function.clone(), d1.pc, d1.belongs_to_var.clone()))
                {
                    for d4 in incoming {
                        let instructions = &program
                            .functions
                            .iter()
                            .find(|x| x.name == d4.function)
                            .context("Cannot find function")?
                            .instructions;

                        let ret_vals = self.return_val(
                            &d4.function,
                            &d2.function,
                            d4.pc,
                            d2.pc,
                            &instructions,
                            graph,
                        )?;

                        let return_site_facts = graph
                            .get_facts_at(&d4.function, d4.pc + 1)?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        for calling_edge in ret_vals.into_iter() {
                            let return_site_fact_of_caller = return_site_facts
                                .iter()
                                .find(|x| {
                                    x.belongs_to_var == calling_edge.get_from().belongs_to_var
                                })
                                .context("Cannot find return site of caller")?;

                            if summary_edge
                                .iter()
                                .find(|x| {
                                    x.get_from() != d4 && x.to() != return_site_fact_of_caller
                                }) //d5 muss hier im test sein
                                .is_none()
                            {
                                summary_edge.push(Edge::Summary {
                                    from: d4.clone(),
                                    to: return_site_fact_of_caller.clone(),
                                });

                                let edges: Vec<_> = path_edge
                                    .iter()
                                    .filter(|x| {
                                        x.to() == d4 && &x.get_from().function == &d4.function
                                    })
                                    .cloned()
                                    .collect();

                                for d3 in edges.into_iter() {
                                    self.propagate(
                                        path_edge,
                                        worklist,
                                        Edge::Path {
                                            from: d3.get_from().clone(),
                                            to: return_site_fact_of_caller.clone(),
                                        },
                                    )?;
                                }
                            }
                        }
                    }
                }
            }
        }

        graph.edges.extend_from_slice(&path_edge);
        graph.edges.extend_from_slice(&summary_edge);

        Ok(())
    }
}
