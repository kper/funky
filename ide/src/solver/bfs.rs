use crate::icfg::graph2::Graph;
use crate::solver::*;
use std::collections::VecDeque;

use log::debug;

pub struct BFS;

impl GraphReachability for BFS {
    fn all_sinks(&mut self, graph: &mut Graph, req: Request) -> Vec<Taint> {
        let mut queue: VecDeque<usize> = VecDeque::new();
        let mut seen = Vec::new();

        if let Some(start) = graph.query(&req) {
            debug!("Adding start {:#?}", start);
            queue.push_back(start.id);
        } else {
            return Vec::new();
        }

        while let Some(node) = queue.pop_front() {
            debug!("Popping node {}", node);
            seen.push(node);
            for child in graph.get_neighbours(node) {
                debug!("Adding child {:#?}", graph.query_by_fact_id(child));

                if !seen.contains(&child) {
                    debug!("queue child {}", child);
                    queue.push_back(child);
                }
            }
        }

        seen.into_iter()
            .map(|x| graph.query_by_fact_id(x).unwrap())
            .map(|x| Taint {
                from: req.variable.clone(),
                from_function: req.function.clone(),
                from_pc: req.pc,
                to: x.belongs_to_var.clone(),
                to_pc: x.pc,
                to_function: x.function.clone(),
            })
            .collect()
    }
}
