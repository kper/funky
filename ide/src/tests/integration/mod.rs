use crate::ir::wasm_ast::IR;
use funky::engine::module::ModuleInstance;
use funky::engine::*;
use insta::assert_snapshot;
use validation::validate;
use wasm_parser::{parse, read_wasm};

use crate::grammar::*;
use crate::icfg::convert2::ConvertSummary;
use crate::icfg::tikz::render_to;
use crate::solver::Request;

macro_rules! wasm {
    ($input:expr) => {{
        // Read it
        let file = read_wasm!($input);
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let imports = Vec::new();

        let instance = ModuleInstance::new(&module);
        let engine = Engine::new(
            instance,
            &module,
            Box::new(funky::debugger::RelativeProgramCounter::default()),
            &imports,
        )
        .unwrap();

        let mut ir = IR::new();

        ir.visit(&engine).unwrap();

        ir
    }};
}

macro_rules! run {
    ($name:expr, $req:expr, $fs:expr) => {
        let ir = wasm!($fs);

        let ir_code = ir.buffer();

        let mut convert = ConvertSummary::new();

        let prog = ProgramParser::new().parse(&ir_code).unwrap();

        let graph = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph);

        assert_snapshot!(format!("{}", $name), output);
    };
}

#[test]
fn test_add() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("add", req, "../tests/add.wasm");
}

#[test]
fn test_fib() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("fib", req, "../tests/fib.wasm");
}

#[test]
fn test_fac() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("fac", req, "../tests/fac.wasm");
}

#[test]
fn test_br_table() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("br_table", req, "../tests/br_table.wasm");
}
