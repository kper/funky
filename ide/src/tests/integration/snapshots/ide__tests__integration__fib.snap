---
source: src/tests/integration/mod.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:\%1] (18) at (2, 4) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (16) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:\%1] (14) at (2, 3) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (12) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:\%1] (10) at (2, 2) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (8) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (4) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (0) at (0, 0) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Block(7)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Assign(1, 0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Call(0, [1], [2])};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Return([2])};
\node[circle,fill,inner sep=1pt,label=left:taut] (951) at (5, 49) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (932) at (5, 48) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (913) at (5, 47) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (894) at (5, 46) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (875) at (5, 45) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (856) at (5, 44) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (837) at (5, 43) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (818) at (5, 42) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (799) at (5, 41) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (419) at (5, 21) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (400) at (5, 20) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (381) at (5, 19) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (362) at (5, 18) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (343) at (5, 17) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (305) at (5, 15) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (286) at (5, 14) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (267) at (5, 13) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (248) at (5, 12) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (229) at (5, 11) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (210) at (5, 10) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (191) at (5, 9) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (172) at (5, 8) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (153) at (5, 7) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (134) at (5, 6) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (115) at (5, 5) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (96) at (5, 4) { };
\node[circle,fill,inner sep=1pt,label=left:\%1] (79) at (7, 3) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (77) at (5, 3) { };
\node[circle,fill,inner sep=1pt,label=left:\%1] (60) at (7, 2) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (58) at (5, 2) { };
\node[circle,fill,inner sep=1pt,label=left:\%1] (41) at (7, 1) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (39) at (5, 1) { };
\node[circle,fill,inner sep=1pt,label=left:\%1] (22) at (7, 0) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (20) at (5, 0) { };
\node[font=\tiny] (note_4) at (3.5, 0.5) {Block(0)};
\node[font=\tiny] (note_5) at (3.5, 1.5) {Block(1)};
\node[font=\tiny] (note_6) at (3.5, 2.5) {Block(3)};
\node[font=\tiny] (note_7) at (3.5, 3.5) {Assign(1, 0)};
\node[font=\tiny] (note_8) at (3.5, 4.5) {Const(2, 1.0)};
\node[font=\tiny] (note_9) at (3.5, 5.5) {BinOp(3, 2, 1)};
\node[font=\tiny] (note_10) at (3.5, 6.5) {Conditional(3, [5, 6])};
\node[font=\tiny] (note_11) at (3.5, 7.5) {Block(5)};
\node[font=\tiny] (note_12) at (3.5, 8.5) {Assign(4, 0)};
\node[font=\tiny] (note_13) at (3.5, 9.5) {Const(5, 1.0)};
\node[font=\tiny] (note_14) at (3.5, 10.5) {BinOp(6, 5, 4)};
\node[font=\tiny] (note_15) at (3.5, 11.5) {Conditional(6, [4])};
\node[font=\tiny] (note_16) at (3.5, 12.5) {Kill(6)};
\node[font=\tiny] (note_17) at (3.5, 13.5) {Kill(5)};
\node[font=\tiny] (note_18) at (3.5, 14.5) {Kill(4)};
\node[font=\tiny] (note_19) at (3.5, 15.5) {Jump(2)};
\node[font=\tiny] (note_20) at (3.5, 16.5) {Jump(6)};
\node[font=\tiny] (note_21) at (3.5, 17.5) {Block(6)};
\node[font=\tiny] (note_22) at (3.5, 18.5) {Assign(7, 0)};
\node[font=\tiny] (note_23) at (3.5, 19.5) {Const(8, -1.0)};
\node[font=\tiny] (note_24) at (3.5, 20.5) {BinOp(9, 8, 7)};
\node[font=\tiny] (note_25) at (3.5, 21.5) {Call(0, [9], [10])};
\node[font=\tiny] (note_26) at (3.5, 22.5) {Assign(11, 0)};
\node[font=\tiny] (note_27) at (3.5, 23.5) {Const(12, -2.0)};
\node[font=\tiny] (note_28) at (3.5, 24.5) {BinOp(13, 12, 11)};
\node[font=\tiny] (note_29) at (3.5, 25.5) {Call(0, [13], [14])};
\node[font=\tiny] (note_30) at (3.5, 26.5) {BinOp(15, 14, 13)};
\node[font=\tiny] (note_31) at (3.5, 27.5) {Kill(15)};
\node[font=\tiny] (note_32) at (3.5, 28.5) {Kill(14)};
\node[font=\tiny] (note_33) at (3.5, 29.5) {Kill(13)};
\node[font=\tiny] (note_34) at (3.5, 30.5) {Kill(12)};
\node[font=\tiny] (note_35) at (3.5, 31.5) {Kill(11)};
\node[font=\tiny] (note_36) at (3.5, 32.5) {Kill(10)};
\node[font=\tiny] (note_37) at (3.5, 33.5) {Kill(9)};
\node[font=\tiny] (note_38) at (3.5, 34.5) {Kill(8)};
\node[font=\tiny] (note_39) at (3.5, 35.5) {Kill(7)};
\node[font=\tiny] (note_40) at (3.5, 36.5) {Kill(3)};
\node[font=\tiny] (note_41) at (3.5, 37.5) {Kill(2)};
\node[font=\tiny] (note_42) at (3.5, 38.5) {Kill(1)};
\node[font=\tiny] (note_43) at (3.5, 39.5) {Return([0])};
\node[font=\tiny] (note_44) at (3.5, 40.5) {Jump(4)};
\node[font=\tiny] (note_45) at (3.5, 41.5) {Block(4)};
\node[font=\tiny] (note_46) at (3.5, 42.5) {Const(16, 0.0)};
\node[font=\tiny] (note_47) at (3.5, 43.5) {Assign(0, 16)};
\node[font=\tiny] (note_48) at (3.5, 44.5) {Kill(16)};
\node[font=\tiny] (note_49) at (3.5, 45.5) {Jump(2)};
\node[font=\tiny] (note_50) at (3.5, 46.5) {Block(2)};
\node[font=\tiny] (note_51) at (3.5, 47.5) {Assign(17, 0)};
\node[font=\tiny] (note_52) at (3.5, 48.5) {Return([17])};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->] (0) edge (4);
		\path[->, blue] (0) [bend right] edge  node { } (4);
		\path[->] (4) edge (10);
		\path[->, blue] (4) [bend right] edge  node { } (10);
		\path[->, blue] (0) [bend right] edge  node { } (8);
		\path[->, blue] (22) [loop right] edge  node { } (22);
		\path[->, blue] (0) [bend right] edge  node { } (14);
		\path[->, blue] (20) [loop right] edge  node { } (20);
		\path[->, blue] (0) [bend right] edge  node { } (12);
		\path[->, blue] (22) [bend right] edge  node { } (41);
		\path[->, blue] (0) [bend right] edge  node { } (18);
		\path[->, blue] (20) [bend right] edge  node { } (39);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (22) [bend right] edge  node { } (60);
		\path[->, blue] (20) [bend right] edge  node { } (58);
		\path[->, blue] (22) [bend right] edge  node { } (79);
		\path[->, blue] (20) [bend right] edge  node { } (77);
		\path[->, blue] (20) [bend right] edge  node { } (96);
		\path[->, blue] (20) [bend right] edge  node { } (115);
		\path[->, blue] (20) [bend right] edge  node { } (134);
		\path[->, blue] (20) [bend right] edge  node { } (153);
		\path[->, blue] (20) [bend right] edge  node { } (343);
		\path[->, blue] (20) [bend right] edge  node { } (172);
		\path[->, blue] (20) [bend right] edge  node { } (362);
		\path[->, blue] (20) [bend right] edge  node { } (191);
		\path[->, blue] (20) [bend right] edge  node { } (381);
		\path[->, blue] (20) [bend right] edge  node { } (210);
		\path[->, blue] (20) [bend right] edge  node { } (400);
		\path[->, blue] (20) [bend right] edge  node { } (229);
		\path[->, blue] (20) [bend right] edge  node { } (419);
		\path[->, blue] (20) [bend right] edge  node { } (799);
		\path[->, blue] (20) [bend right] edge  node { } (248);
		\path[->, blue] (20) [bend right] edge  node { } (818);
		\path[->, blue] (20) [bend right] edge  node { } (267);
		\path[->, blue] (20) [bend right] edge  node { } (837);
		\path[->, blue] (20) [bend right] edge  node { } (286);
		\path[->, blue] (20) [bend right] edge  node { } (856);
		\path[->, blue] (20) [bend right] edge  node { } (305);
		\path[->, blue] (20) [bend right] edge  node { } (875);
		\path[->, blue] (20) [bend right] edge  node { } (894);
		\path[->, blue] (20) [bend right] edge  node { } (913);
		\path[->, blue] (20) [bend right] edge  node { } (932);
		\path[->, blue] (20) [bend right] edge  node { } (951);


                \end{tikzpicture}
            \end{document}
